### PR TITLE
Make "applyToParams" a Future + add "this." when decoding JSON

### PR DESCRIPTION
**Pull request by noordawod**
This pr brings back 2 lost "features" that were present earlier (and of course with the proper explanation as to why we need them back), and introduces a couple of new features as well:

1. Inside a function, private-style variables (`var _json`) make no sense – so removed the underscore.
2. ~~The `authNames` were mysteriously dropped, why? These are important information to each endpoint, and the authentication implementation may need to check which authentications to apply for each operation.~~
3. Removed use of the bang operator (`!`) when activating the authentication implementation – just use what Dart gives us out of the box: `authentication?.applyToParams(…)`
4. Converted the `applyToParams` method to a `Future<void>`. Custom implementations may need to retrieve data from persistent stores, for example, and so may need this.
5. ~~Since `Authentication` is an interface, and in case `ApiClient` is inherited, one can now declare the generic `T extends Authentication` class upon extension. Helps to access properties added in the inherited class (yes, this is backward-compatible and Dart, luckily, doesn't scream about it.)~~

### PR checklist
 
- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [X] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (6.1.0) (minor release - breaking changes with fallbacks), `7.0.x` (breaking changes without fallbacks)
- [X] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@jimschubert @dr4ke616 @karismann @Zomzog @andrewemery @4brunu @yutaka0m @wing328 @jaumard @josh-burton @amondnet @sbu-WBT @kuhnroyal @agilob @ahmednfwela

### CODE REVIEW

File: modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache [hunk 1]
```diff
@@ -52,7 +52,7 @@ class {{{classname}}} {
   String toString() => '{{{classname}}}[{{#vars}}{{{name}}}=${{{name}}}{{^-last}}, {{/-last}}{{/vars}}]';
 
   Map<String, dynamic> toJson() {
-    final _json = <String, dynamic>{};
+    final json = <String, dynamic>{};
```
Comment (wing328) [inline modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:55]: @noordawod thanks for the PR. What if one of the property names is `json`? Would that still work?

Comment (noordawod) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:55]: Of course, why wouldn't it. This is a language-level variable name, while a `"json"` is just a key in a Map. The same question can be asked even the variable name is `_json` – there's no difference really.

Bottom line: Two different concepts and they don't mix, no problem.

**Comment by ahmednfwela** [comment]
I remember this being changed due to https://github.com/OpenAPITools/openapi-generator/pull/11360 , as a fix for https://github.com/OpenAPITools/openapi-generator/issues/7057
will this issue regress after this PR ?

File: modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache (hunk 1 above)
Comment (ahmednfwela) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:55]: I think changing property access in `toJson` to be `this.member` instead of just `member` can also avoid any issues

Comment (wing328) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:55]: In https://github.com/OpenAPITools/openapi-generator/pull/13120/files#diff-19b5d9bf7f13b303faf741f31a0d3853aa461419c6e9fd97128328459f592ad3R58:

```
    if ({{{name}}} != null) {
```

Note that it's not using `baseName` (which can have a leading underscore).

so that line will become 
```
    if (json != null) {
```
if one of the property names is `json`

Would `json` refers to the property or the `final json` declared earlier in the function?

Comment (noordawod) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:55]: A fair and valid point. I guess we can just use `this.` to reference the properties, which should fix any disambiguation. Thoughts?

**Comment by noordawod** [comment]
> I remember this being changed due to #11360 , as a fix for #7057 will this issue regress after this PR ?

Hmm, I didn't know about the openapi requirement. It makes sense, so I will revert back the changes but keep the `Future` nature of `applyToParams`, which is okay I think.

**Comment by ahmednfwela** [comment]
yes, I think making it a Future is a good improvement, especially since it's a common scenario to persist and read auth tokens from local storage, which is usually a Future

File: modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache (hunk 1 above)
Comment (wing328) [reply modules/openapi-generator/src/main/resources/dart2/serialization/native/native_class.mustache:55]: I'm ok with any approach as long as the output (auto-generated code) works without issues.

(I"m not a Dart developer so I do not know which approach works better from a Dart developer's perspective.)

File: modules/openapi-generator/src/main/resources/dart2/api_client.mustache [hunk 2]
```diff
@@ -1,11 +1,13 @@
 {{>header}}
 {{>part_of}}
-class ApiClient {
-  ApiClient({this.basePath = '{{{basePath}}}', this.authentication});
+class ApiClient<T extends Authentication> {
```
Comment (ahmednfwela) [inline modules/openapi-generator/src/main/resources/dart2/api_client.mustache:3]: Why is this change necessary? even if the developer inherits `ApiClient`, there is no real use for the generic parameter

File: modules/openapi-generator/src/main/resources/dart2/api_client.mustache [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
+  ApiClient({this.basePath = '{{{basePath}}}', this.authentication,});
 
   final String basePath;
+  final T? authentication;
 
   var _client = Client();
+  final _defaultHeaderMap = <String, String>{};
```
Comment (ahmednfwela) [inline modules/openapi-generator/src/main/resources/dart2/api_client.mustache:10]: if you want to expose this, shouldn't just a simple rename to
```dart
final defaultHeaderMap = <String, String>{};
``` 
work ?

File: modules/openapi-generator/src/main/resources/dart2/api_client.mustache (hunk 2 above)
Comment (noordawod) [reply modules/openapi-generator/src/main/resources/dart2/api_client.mustache:3]:
```dart
abstract CustomAuthenticationMechanism extends Authentication {
  Future<String?> get webTokenProvider;
}

class CustomApiClient extends ApiClient<CustomAuthenticationMechanism> {
  // We have access to authentication.webTokenProvider
}
```

Comment (ahmednfwela) [reply modules/openapi-generator/src/main/resources/dart2/api_client.mustache:3]: why not :
```dart
class CustomApiClient extends ApiClient {
  @override
  final CustomAuthenticationMechanism? authentication;
}
```

File: modules/openapi-generator/src/main/resources/dart2/api_client.mustache (hunk 3 above)
Comment (noordawod) [reply modules/openapi-generator/src/main/resources/dart2/api_client.mustache:10]: I just moved properties around so private ones huddle together, so no logic change in this pr.

File: modules/openapi-generator/src/main/resources/dart2/api_client.mustache (hunk 2 above)
Comment (noordawod) [reply modules/openapi-generator/src/main/resources/dart2/api_client.mustache:3]: Looks promising, does it work? 🙂 I'll check it and if works then I'll revert. Thanks!

Comment (noordawod) [reply modules/openapi-generator/src/main/resources/dart2/api_client.mustache]: Looks like it works, so reverted addition of `T` generic. Thanks for the tip!

**Comment by wing328** [comment]
Any one has more feedback on this change?

cc @jaumard (2018/09) @josh-burton (2019/12) @amondnet (2019/12) @sbu-WBT (2020/12) @kuhnroyal (2020/12) @agilob (2020/12) @ahmednfwela (2021/08)

**Review by ahmednfwela — APPROVED** [review]

**Comment by ahmednfwela** [comment]
LGTM

**Comment by jld3103** [comment]
Hello, I just upgraded OpenAPI generator and I'm not at all happy with applyToParams being a Future. It breaks my whole application and just doesn't make sense. When a custom implementation is used, the credentials should just be passed to the class instead of being fetched inside the class. Even the name of the function already means that it only applies and not that it fetches the credentials. Sorry, but this change has to be reverted. The uses case of this is invalid as it means the architecture of the applications is broken. The function for applying the credentials to the requests should never be responsible for fetching the credentials. This design also means the credentials could be fetched on every request.

**Comment by noordawod** [comment]
"It breaks my whole application and just doesn't make sense." -- made sense to us all ☺️

"When a custom implementation is used, the credentials should just be passed to the class instead of being fetched inside the class" -- that's one way to do it, which can be done with `applyToParams` being a Future.

"The uses case of this is invalid as it means the architecture of the applications is broken." -- could it be that your app's architectural design choices are wrong?

"This design also means the credentials could be fetched on every request." -- makes perfect sense to me; fresh credentials from the source of truth.

Would you like to share your use case? Maybe we can assist you.

**Comment by jld3103** [comment]
sorry i just had to rant a little. i don't want to bother anymore with openapi generator since it breaks so often. this is a breaking change and it has been merged into a minor version. this is inacceptable, but i don't care anymore.

**Comment by ahmednfwela** [comment]
@jld3103 can you share the specific code that broke by upgrading this version?
it would be helpful if you could open a new issue
